pkg/backup/compaction_job.go

```diff
@@ -525,17 +525,57 @@ func updateCompactionBackupDetails(
 	return compactedDetails, nil
 }
 
-// compactIntroducedSpans takes a compacted backup manifest and the full chain of backups it belongs
-// to and computes the introduced spans for the compacted backup.
-func compactIntroducedSpans(
-	ctx context.Context, manifest backuppb.BackupManifest, chain compactionChain,
-) (roachpb.Spans, error) {
-	if err := checkCoverage(ctx, manifest.Spans, chain.backupChain); err != nil {
+// createCompactionManifest creates a new manifest for a compaction job and its
+// compacted chain. The details should have its targets resolved.
+func (c compactionChain) createCompactionManifest(
+	ctx context.Context, details jobspb.BackupDetails,
+) (*backuppb.BackupManifest, error) {
+	// TODO (kev-cao): Will need to update the SSTSinkKeyWriter to support
+	// range keys.
+	lastBackup := c.lastBackup()
+	if len(lastBackup.Tenants) != 0 {
+		return nil, errors.New("backup compactions does not yet support range keys")
+	}
+	introducedSpans, err := c.compactIntroducedSpans(ctx)
+	if err != nil {
+		return nil, err
+	}
+	cManifest := lastBackup
+	cManifest.StartTime = c.start
+	cManifest.EndTime = c.end
+	cManifest.Descriptors = details.ResolvedTargets
+	cManifest.CompleteDbs = details.ResolvedCompleteDbs
+	cManifest.IntroducedSpans = introducedSpans
+	cManifest.IsCompacted = true
+
+	cManifest.Dir = cloudpb.ExternalStorage{}
+	// As this manifest will be used prior to the manifest actually being written
+	// (and therefore its external files being created) we first set
+	// HasExternalManifestSSTs to false and allow it to be set to true later when
+	// it is written to storage.
+	cManifest.HasExternalManifestSSTs = false
+	// While we do not compaction on revision history backups, we still need to
+	// nil out DescriptorChanges to avoid the placeholder descriptor from the
+	// manifest of the last incremental.
+	cManifest.DescriptorChanges = nil
+	cManifest.Files = nil
+	cManifest.EntryCounts = roachpb.RowCount{}
```

Comment:
This logic is becoming complex enough that I think we need to add a specific unit test along the lines of:

TestCheckCompactedManifestFields:

- asserts which fields are inherited
- asserts which fields are nil'd out
- asserts which fields come from details (i'm unsure if those can simply get inherited)
- fails if an engineer adds a new manifest field and has not modified this test

let me know if you want to discuss how to write such a test.

Reply:
Yeah that sounds like a good idea to me.

> fails if an engineer adds a new manifest field and has not modified this test

This sounds like a fun test to write so I'll give it a shot first